### PR TITLE
Update to GTK4 and drop X11 dependency

### PR DESCRIPTION
**Pull request by kanru**

### CODE REVIEW

**Comment by kanru** [comment]
There's one remaining issue: Shift + key somehow does not produce a upper case letter in English mode.

File: src/ibus-chewing-engine.c [hunk 1]
```diff
@@ -112,9 +104,11 @@ extern gint ibus_chewing_verbose;
 typedef IBusChewingEngine Self;
 typedef IBusChewingEngineClass SelfClass;
 
+G_DEFINE_TYPE_WITH_CODE(IBusChewingEngine, ibus_chewing_engine, IBUS_TYPE_ENGINE, G_ADD_PRIVATE(IBusChewingEngine));
+
 /* here are local prototypes */
-static void ibus_chewing_engine_init (IBusChewingEngine * self) G_GNUC_UNUSED;
-static void ibus_chewing_engine_class_init (IBusChewingEngineClass * klass) G_GNUC_UNUSED;
+// static void ibus_chewing_engine_init (IBusChewingEngine * self) G_GNUC_UNUSED;
+// static void ibus_chewing_engine_class_init (IBusChewingEngineClass * klass) G_GNUC_UNUSED;
```
Comment (kanru) [inline src/ibus-chewing-engine.c]: Remove

File: src/ibus-chewing-engine.c [hunk 2]
```diff
@@ -223,16 +200,15 @@ static void
 ibus_chewing_engine_init (IBusChewingEngine * self G_GNUC_UNUSED)
 {
 #define __GOB_FUNCTION__ "IBus:Chewing:Engine::init"
-	self->_priv = G_TYPE_INSTANCE_GET_PRIVATE(self,IBUS_TYPE_CHEWING_ENGINE,IBusChewingEnginePrivate);
+	// ibus_chewing_engine_get_instance_private (self) = G_TYPE_INSTANCE_GET_PRIVATE(self,IBUS_TYPE_CHEWING_ENGINE,IBusChewingEnginePrivate);
```
Comment (kanru) [inline src/ibus-chewing-engine.c]: Remove

File: src/ibus-chewing-engine.c [hunk 3]
```diff
@@ -346,7 +305,7 @@ ibus_chewing_engine_class_init (IBusChewingEngineClass * klass G_GNUC_UNUSED)
 	GObjectClass *g_object_class G_GNUC_UNUSED = (GObjectClass*) klass;
 	IBusEngineClass *ibus_engine_class = (IBusEngineClass *)klass;
 
-	g_type_class_add_private(klass,sizeof(IBusChewingEnginePrivate));
+	// g_type_class_add_private(klass,sizeof(IBusChewingEnginePrivate));
```
Comment (kanru) [inline src/ibus-chewing-engine.c]: Remove

File: src/ibus-chewing-engine.c [hunk 4]
```diff
@@ -468,18 +427,18 @@ ibus_chewing_engine_restore_mode (IBusChewingEngine * self)
 	g_return_if_fail (IBUS_IS_CHEWING_ENGINE (self));
 {
 	
-        IBUS_CHEWING_LOG(DEBUG, "restore_mode() statusFlags=%x", self->_priv->statusFlags);
-        if (self->_priv->pDisplay != NULL) {
-            /* Restore Led Mode only make sense if pDisplay is available */
+        IBUS_CHEWING_LOG(DEBUG, "restore_mode() statusFlags=%x", selfp->statusFlags);
+        GdkDisplay *display = gdk_display_get_default();
+        if (display != NULL) {
+            GdkSeat *seat = gdk_display_get_default_seat(display);
```
Comment (kanru) [inline src/ibus-chewing-engine.c:429]: `gdk_display_get_default_seat` may return NULL

File: src/ibus-chewing-engine.c [hunk 5]
```diff
@@ -468,18 +427,18 @@ ibus_chewing_engine_restore_mode (IBusChewingEngine * self)
...
+            GdkDevice *keyboard = gdk_seat_get_keyboard(seat);
```
Comment (kanru) [inline src/ibus-chewing-engine.c]: `gdk_seat_get_keyboard` also may return NULL

Fix those I think we can then try to run the test in weston.

**Comment by yan12125** [comment]
Is it intended to remove many icons? I tried current git-master (4f8f25c) and found icons are missing in several places.

**Comment by kanru** [comment]
> Is it intended to remove many icons? I tried current git-master ([4f8f25c](https://github.com/chewing/ibus-chewing/commit/4f8f25c05070c3b37cd762fe5f8a3c19b18c93e3)) and found icons are missing in several places.

你是說 systray icon 嗎？那應該是唯一會用到 icon 的地方

**Comment by kanru** [comment]
> > Is it intended to remove many icons? I tried current git-master ([4f8f25c](https://github.com/chewing/ibus-chewing/commit/4f8f25c05070c3b37cd762fe5f8a3c19b18c93e3)) and found icons are missing in several places.
> 
> 你是說 systray icon 嗎？那應該是唯一會用到 icon 的地方

Desktop file 的 icon 我有留下來，另外我也有計畫把 systray 加回來，畢竟有些 DE 沒有直接的 ibus integration

**Comment by yan12125** [comment]
> > Is it intended to remove many icons? I tried current git-master ([4f8f25c](https://github.com/chewing/ibus-chewing/commit/4f8f25c05070c3b37cd762fe5f8a3c19b18c93e3)) and found icons are missing in several places.
> 
> 你是說 systray icon 嗎？那應該是唯一會用到 icon 的地方

「IBus 偏好設定」中的icon也不見了，我提了一個新 issue: https://github.com/chewing/ibus-chewing/issues/181

**Pull request merged**